qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java

```diff
@@ -0,0 +1,111 @@
+ * products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
+ * NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.qulice.checkstyle;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck;
+import java.io.File;
+import java.util.List;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Make sure each line indentation is either:
+ * <ul>
+ * <li>the same as previous one or less
+ * <li>bigger than previous by exactly 4
+ * </ul>
+ * All other cases must cause a failure.
+ *
+ * @author Hamdi Douss (douss.hamdi@gmail.com)
+ * @version $Id$
+ */
+public final class QuliceIndentationCheck extends AbstractFileSetCheck {
+    /**
+     * {@inheritDoc}
+     * @checkstyle MagicNumberCheck (13 lines)
+     */
+    @Override
+    public void processFiltered(final File file, final List<String> lines) {
+        int previous = 0;
+        for (int pos = 0; pos < lines.size(); pos += 1) {
+            final String line = lines.get(pos);
+            final int current = indentation(line);
+            if (inCommentBlock(line) || line.length() == 0) {
+                continue;
+            }
+            if (current > previous
+                && current != previous + 4) {
+                this.log(
+                    pos + 1,
+                    String.format(
+                        StringUtils.join(
+                            "Indentation (%d) must be same or ",
+                            "less than previous line (%d), or ",
+                            "bigger by exactly 4"
+                        ),
+                        current,
+                        previous
+                    )
+                );
+            }
+            previous = current;
+        }
+    }
+
+    /**
+     * Checks if the line belongs to a comment block.
+     * @param line Input.
+     * @return True if the line belongs to a comment block.
+     */
+    private static boolean inCommentBlock(final String line) {
+        final String trimmed = line.trim();
+        return trimmed.length() > 0
+            && (trimmed.charAt(0) == '*'
+                || trimmed.startsWith("/*")
+                || trimmed.startsWith("*/")
+                );
```

Comment:
@pinaf Unlike lines in comment block, single line comments should follow the rule. The rule is even forced by Checkstyle.